src/Hpack/Config.hs

```diff
@@ -691,37 +729,55 @@ toLibrary dir name globalOptions library = traverse fromLibrarySection sect >>=
+    updateConditionals :: Section LibrarySection LibrarySection -> Section LibrarySection Library -> Section Library Library
+    updateConditionals source target = target { sectionConditionals = (fmap . fmap) fromLibrarySectionPlain (sectionConditionals source) }
+
+    fromLibrarySectionPlain :: LibrarySection -> Library
+    fromLibrarySectionPlain LibrarySection{..} = Library librarySectionExposed (fromMaybeList librarySectionExposedModules) (fromMaybeList librarySectionOtherModules) (fromMaybeList librarySectionReexportedModules)
+
+toExecutables :: FilePath -> String -> Section Empty global -> [(String, Section ExecutableSection ExecutableSection)] -> IO ([String], [Section Executable Executable])
 toExecutables dir packageName_ globalOptions executables = do
   result <- mapM toExecutable sections >>= mapM (expandForeignSources dir)
   let (warnings, xs) = unzip result
   return (concat warnings, xs)
   where
-    sections :: [(String, Section ExecutableSection)]
-    sections = map (fmap $ mergeSections globalOptions) executables
+    sections :: [(String, Section ExecutableSection ExecutableSection)]
+    sections = map (fmap $ mergeSections emptyExecutable globalOptions) executables
+
+    emptyExecutable :: ExecutableSection
+    emptyExecutable = ExecutableSection Nothing Nothing
 
-    toExecutable :: (String, Section ExecutableSection) -> IO (Section Executable)
+    toExecutable :: (String, Section ExecutableSection ExecutableSection) -> IO (Section Executable Executable)
     toExecutable (name, sect@Section{..}) = do
       (executable, ghcOptions) <- fromExecutableSection sectionData
-      return (sect {sectionData = executable, sectionGhcOptions = sectionGhcOptions ++ ghcOptions})
+      return (sect
+        { sectionData = executable
+        , sectionGhcOptions = sectionGhcOptions ++ ghcOptions
+        , sectionConditionals = (fmap . fmap) fromExecutableSectionNamed sectionConditionals
+        })
       where
         fromExecutableSection :: ExecutableSection -> IO (Executable, [GhcOption])
         fromExecutableSection ExecutableSection{..} = do
           modules <- maybe inferModules (return . fromList) executableSectionOtherModules
-          return (Executable name mainSrcFile modules, ghcOptions)
+          return (Executable (Just name) mainSrcFile modules, ghcOptions)
           where
             inferModules :: IO [String]
             inferModules = filterMain . (++ [pathsModule]) . concat <$> mapM (getModules dir) sectionSourceDirs
 
             pathsModule = pathsModuleFromPackageName packageName_
 
             filterMain :: [String] -> [String]
-            filterMain = maybe id (filter . (/=)) (toModule $ splitDirectories executableSectionMain)
+            filterMain = maybe (const []) (\m -> maybe id (filter . (/=)) (toModule $ splitDirectories m)) executableSectionMain
```

Comment:
I think `const []` needs to be `id` instead (if there is no `main` we don't filter anything).